File: src/types/vuln.ts
import { z } from "zod";
import { pluginSiteSchema } from "./plugin";

export const cweSchema = z.object({
  cwe: z.string(),
  name: z.string(),
  description: z.string(),
});

export const cvssSchema = z.object({
  version: z.string(),
  vector: z.string(),
  av: z.string(),
  ac: z.string(),
  pr: z.string(),
  ui: z.string(),
  s: z.string(),
  c: z.string(),
  i: z.string(),
  a: z.string(),
  score: z.string(),
  severity: z.string(),
  exploitable: z.string(),
  impact: z.string().nullable(),
});

export const impactSchema = z.object({
  cvss: cvssSchema.optional(),
  cwe: z.array(cweSchema).optional(),
});

export const sourceSchema = z.object({
  id: z.string(),
  name: z.string(),
  link: z.string(),
  description: z.string().nullable(),
  date: z.string().nullable(),
});

export const operatorSchema = z.object({
  min_version: z.string().nullable(),
  min_operator: z.string().nullable(),
  max_version: z.string().nullable(),
  max_operator: z.string().nullable(),
  unfixed: z.string(),
  closed: z.string(),
});

export const vulnerabilitySchema = z.object({
  uuid: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  operator: operatorSchema,
  source: z.array(sourceSchema),
  impact: z.union([impactSchema, z.array(z.never())]),
});

export const vulnDataSchema = z.object({
  name: z.string().nullable(),
  plugin: z.string(),
  link: z.string().nullable(),
  latest: z.string().nullable(),
  vulnerability: z.array(vulnerabilitySchema).nullable(),
});

export const vulnResponseSchema = z.object({
  error: z.number().default(1),
  message: z.string().default("").nullable(),
  data: vulnDataSchema.optional(),
  updated: z.coerce.number().default(0),
});

export type Cwe = z.infer<typeof cweSchema>;
export type Cvss = z.infer<typeof cvssSchema>;
export type Impact = z.infer<typeof impactSchema>;
export type Source = z.infer<typeof sourceSchema>;
export type Operator = z.infer<typeof operatorSchema>;
export type Vulnerability = z.infer<typeof vulnerabilitySchema>;
export type VulnData = z.infer<typeof vulnDataSchema>;
export type VulnResponse = z.infer<typeof vulnResponseSchema>;

export const vulnReportSchema = z.object({
  plugin: z.string(),
  vulnerability: vulnerabilitySchema,
  sites: z.array(pluginSiteSchema),
});

export type VulnReport = z.infer<typeof vulnReportSchema>;

File: internal/models/vuln.go
package models

import "encoding/json"

type Cwe struct {
	Cwe         string `json:"cwe"`
	Name        string `json:"name"`
	Description string `json:"description"`
}

type Cvss struct {
	Version     string  `json:"version"`
	Vector      string  `json:"vector"`
	Av          string  `json:"av"`
	Ac          string  `json:"ac"`
	Pr          string  `json:"pr"`
	Ui          string  `json:"ui"`
	S           string  `json:"s"`
	C           string  `json:"c"`
	I           string  `json:"i"`
	A           string  `json:"a"`
	Score       string  `json:"score"`
	Severity    string  `json:"severity"`
	Exploitable string  `json:"exploitable"`
	Impact      *string `json:"impact"`
}

type Impact struct {
	Cvss *Cvss `json:"cvss,omitempty"`
	Cwe  []Cwe `json:"cwe,omitempty"`
}

// UnmarshalJSON custom unmarshaler for Impact to handle empty arrays.
func (i *Impact) UnmarshalJSON(data []byte) error {
	if string(data) == "[]" || string(data) == "null" {
		return nil
	}
	type Alias Impact
	aux := (*Alias)(i)
	return json.Unmarshal(data, aux)
}

type Source struct {
	Id          string  `json:"id"`
	Name        string  `json:"name"`
	Link        string  `json:"link"`
	Description *string `json:"description"`
	Date        *string `json:"date"`
}

type Operator struct {
	MinVersion  *string `json:"min_version"`
	MinOperator *string `json:"min_operator"`
	MaxVersion  *string `json:"max_version"`
	MaxOperator *string `json:"max_operator"`
	Unfixed     string  `json:"unfixed"`
	Closed      string  `json:"closed"`
}

type Vulnerability struct {
	Uuid        string       `json:"uuid"`
	Name        string       `json:"name"`
	Description *string      `json:"description"`
	Operator    Operator     `json:"operator"`
	Source      []Source     `json:"source"`
	Impact      *Impact      `json:"impact,omitempty"`
	Sites       []PluginSite `json:"sites,omitempty"`
}

type VulnData struct {
	Name          *string         `json:"name"`
	Plugin        string          `json:"plugin"`
	Link          *string         `json:"link"`
	Latest        *string         `json:"latest"`
	Vulnerability []Vulnerability `json:"vulnerability"`
}

type VulnResponse struct {
	Error   int       `json:"error"`
	Message *string   `json:"message"`
	Data    *VulnData `json:"data,omitempty"`
	Updated any       `json:"updated"`
}

type VulnReport struct {
	Plugin        string        `json:"plugin"`
	Slug          string        `json:"slug"`
	PluginName    string        `json:"plugin_name"`
	Vulnerability Vulnerability `json:"vulnerability"`
	Sites         []PluginSite  `json:"sites"`
}

type VulnPlugin struct {
	PluginName    string          `json:"plugin_name"`
	Version       string          `json:"version"`
	Cvss          *float64        `json:"cvss"`
	Vulnerability []Vulnerability `json:"vulnerability"`
}
